include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_for.h

```diff
@@ -192,8 +192,20 @@ __parallel_for(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&&
     {
         if (__count >= __large_submitter::__estimate_best_start_size(__exec, __brick))
         {
-            return __large_submitter{}(std::forward<_ExecutionPolicy>(__exec), __brick, __count,
-                                       std::forward<_Ranges>(__rngs)...);
+            // Passing begin() of each range is needed for the icpx compiler to vectorize. The indirection introduced
+            // by our all / guard views interfere with compiler vectorization. At this point, we have ensured that
+            // input is contiguous and can be operated on directly. The begin() function for these views will return a
+            // pointer which is passed to the kernel.
+            if constexpr (_Fp::__can_vectorize)
+            {
+                return __large_submitter{}(std::forward<_ExecutionPolicy>(__exec), __brick, __count,
+                                           std::forward<_Ranges>(__rngs).begin()...);
+            }
+            else
+            {
+                return __large_submitter{}(std::forward<_ExecutionPolicy>(__exec), __brick, __count,
+                                           std::forward<_Ranges>(__rngs)...);
+            }
```

Comment:
Is there a need for this `else`?  Is there a disadvantage to passing passing an iterator rather than a range in cases where we cannot vectorize?

Reply:
With the latest commit, ranges over SYCL buffers must take this path as we cannot call `begin()` on the host (it returns a pointer to the first element in the underlying accessor). This means the USM path will likely be faster to the user until this issue can be fully resolved.

Reply:
Ah, yes that makes sense...